NOISSUE - Use `StatusUnauthorized` instead of `StatusForbidden`

sprql opened the pull request:
Use `401 Unauthorized` instead of `403 Forbidden` for errors related to 
authentication. 

According to [https://tools.ietf.org/html/rfc6750#section-3](https://tools.ietf.org/html/rfc6750#section-3)

manuio commented on authn/api/http/endpoint_test.go:
```diff
@@ -219,7 +219,7 @@ func TestRetrieve(t *testing.T) {
 			desc:   "retrieve a key unauthorized",
```
If you have time I would appreciate that you also fix this logs. It should be `retrieve a token unauthorized` or even `retrieve an unauthorized token`

sprql replied:
Ok, I'll do

drasko commented on readers/api/transport.go:
```diff
@@ -28,7 +28,8 @@ const (
 	errInvalidRequest     = errors.New("received invalid request")
-	errUnauthorizedAccess = errors.New("missing or invalid credentials provided")
+	errUnauthorizedAccess = errors.New("missing credentials provided")
+	errAccessDenied       = errors.New("invalid credentials provided")
```
This should be kept consistent all over the code. If there is a difference between missing and invalid credentials (AuthN/AuthZ), then it should be detected in other places, and consequently 401/403 (both of them) codes should be used.

sprql replied:
There are 3 cases: 

- Missing token — 401
- Expired token (or maybe invalid by another reason) — 401
- Valid token, but access to some resource with this token is denied — 403

In this example, the current error message ("invalid credentials provided") maybe is a little bit confusing and should be changed to something else (like "permission denied")

Also maybe I should rename `errAccessDenied` to `errPermissionDenied`

drasko replied:
> There are 3 cases:
>
>Missing token — 401
>Expired token (or maybe invalid by another reason) — 401
>Valid token, but access to some resource with this token is denied — 403

@sprql exactly - well understood. Name of the error does not bother me much - `errAccessDenied` is OK and it relates to `403` rather than `401`, i.e. it is self-documenting enough.

What I wanted to check with you is did you cover all the code code base and identified where one of these 3 cases might appear and well differentiated between them - because here I can see `403` practically appearing just at one place in the whole Mainflux, according to your changes.

I did not review whole PR in details - it's bit difficult to review, this is why I am asking (looks to me you know your way areound Mainflux now and you know what you are doing, which is very good).

Anyway - thanks for this PR, it is very nice and think that you are really on a good track here for a nice improvement. Just please do not forget to update Swagger files as well.

@nmarcetic and @dusanb94 should review this one as well.

dborovcanin replied:
I agree with this approach. This will later translate to handling errors from AuthN and AuthZ services (at the moment, only AuthN exists), where errors from AuthN will cause `Forbidden`, and errors from AuthZ will be `Unauthorized`.

drasko replied:
Great, we have general alignment of maintainers then.

@sprql please double-check the PR, ad final touches prepare it for meging. Please do not forget to add Swagger changes.

After that PR will be merged.

sprql replied:
@drasko, currently, the code mainly uses the `errUnauthorizedAccess`, both for AuthN & AuthZ errors, therefore, we will need to enter a new error type in these places. 

One option is to use error `errUnauthorizedAccess` for AuthN, and enter `errPermissionDenied` for AuthZ. But 'errUnauthorizedAccess' is misleading because of its name. 

Alternatively, we can enter `errUnauthenticated' for AuthN errors (like in gRPC).

In other words, the following options are possible:
`errUnauthorizedAccess` (401), `errPermissionDenied` (403)
or `errUnauthenticated` (401), `errUnauthorizedAccess` (403)
or `errUnauthenticated` (401), `errPermissionDenied` (403)

What do you think about it?

drasko replied:
Good question... I would avoid option 2, as this can be misleading (`errUnauthorizedAccess` might be related to 401 when here should be related to 403).

Between options 1 and 3, probably 3 is the least ambigous, although we could probably understant 1 as well now that we know what this is about...

I would personally vote for 3, to remove any abiguity, but I am also interested what others think. @manuio @nmarcetic @dusanb94 @mteodor @blokovi

manuio replied:
I also vote for 3

sprql replied:
I updated PR using option 3

drasko replied:
@sprql please do ot hate me, but after looking at PR I realized that actually option 2 would be more apropriate for Mainflux.

Actually, in Mainflux we have AuthN and AuthZ services, and I think that we should align errors to these functionalities, and not to outdated and missleading HTTP status names.

So I propose option 2, with a slight change: instad of `errUnauthorizedAccess` for `403` just use name `errUnauthorized`, to align it to it's 401 counterpart (`errUnauthenticated`).

Sorry for this, but I think now that everything is almost done, it will mostly be find-replace job :).

sprql replied:
@drasko I've updated, could you look at this again, I also updated error messages

dborovcanin commented on authn/api/http/transport.go:
```diff
@@ -103,8 +103,8 @@ func encodeError(_ context.Context, err error, w http.ResponseWriter) {
 		w.WriteHeader(http.StatusBadRequest)
-	case errors.Contains(err, authn.ErrUnauthorizedAccess):
-		w.WriteHeader(http.StatusForbidden)
+	case errors.Contains(err, authn.ErrUnauthenticated):
+		w.WriteHeader(http.StatusUnauthorized)
```
Should there be 2 cases (in this particular order):
```go
case errors.Contains(err, authn.ErrUnauthenticated):
    w.WriteHeader(http.StatusUnauthorized)
case errors.Contains(err, authn.ErrUnauthorized):
    w.WriteHeader(http.StatusForbidden)
```
The service itself won't return ErrUnauthenticated, but our JWT implementation (or transport layer in the case `key==""`) might.

dborovcanin commented on authn/swagger.yaml:
```diff
@@ -54,7 +54,7 @@ paths:
         400:
           description: Failed due to malformed query parameters.
-        403:
+        401:
```
If we include both 401 and 403 (see my previous comment), ew should update API docs accordingly.

dborovcanin commented on bootstrap/api/transport.go:
```diff
@@ -289,8 +289,8 @@ func encodeError(_ context.Context, err error, w http.ResponseWriter) {
-		case errors.Contains(errorVal, bootstrap.ErrUnauthorizedAccess):
-			w.WriteHeader(http.StatusForbidden)
+		case errors.Contains(errorVal, bootstrap.ErrUnauthenticated):
+			w.WriteHeader(http.StatusUnauthorized)
```
The same remark from Authn service for adding `Unauthorized` case applies here.
Not to repeat myself and spam this thread, please take a look at all the other services and consider adding both cases both in the transport layer (`svc/api/transport.go`) and API docs (Swagger files).

sprql replied:
> @sprql Please don't address my comments yet. @mainflux/maintainers Need to agree on this. We'll do that internally, bring the final decision and let you know. 

@dusanb94  Have you made a decision on this? Should I add this cases?

dborovcanin replied:
@sprql Let's do it like that. Please update the branch, add these cases, and let's review and merge.

sprql committed:
NOISSUE - Introduce `ErrUnauthenticated` and `ErrUnauthorized` instead of `ErrUnauthorizedAccess`

Use `401 Unauthorized` for errors related to authentication and `403 Forbidden`  for errors related to authorization.

According to https://tools.ietf.org/html/rfc6750#section-3

Signed-off-by: Alexander Obukhov <dev@sprql.space>
